app/src/main/java/org/mozilla/fenix/components/metrics/AppStartupTelemetry.kt

```diff
@@ -0,0 +1,109 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.components.metrics
+
+import android.content.Intent
+import androidx.annotation.VisibleForTesting
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleObserver
+import androidx.lifecycle.OnLifecycleEvent
+import androidx.lifecycle.ProcessLifecycleOwner
+import mozilla.components.support.utils.SafeIntent
+import org.mozilla.fenix.components.metrics.Event.AppStartupSourceAndType
+import org.mozilla.fenix.components.metrics.Event.AppStartupSourceAndType.Source
+import org.mozilla.fenix.components.metrics.Event.AppStartupSourceAndType.Type
+import org.mozilla.fenix.components.metrics.Event.AppStartupSourceAndType.Type.COLD
+import org.mozilla.fenix.components.metrics.Event.AppStartupSourceAndType.Type.WARM
+
+/**
+ * Tracks application startup source, type, and whether or not activity has savedInstance to restore
+ * the activity from. Sample metric = [source = COLD, type = APP_ICON, hasSavedInstance = false].
+ * The basic idea is to collect these metrics from different phases of startup through
+ * [AppStartupSourceAndType]and finally report them on Activity's onResume() function.
+ * If application wasn't backgrounded before onResume() is called, we don't report the metric.
+ */
+class AppStartupTelemetry(private val metrics: MetricController) : LifecycleObserver {
+
+    init {
+        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
+    }
+
+    private var appOpenedAllSourceStartup = AppStartupSourceAndType()
+
+    private var isMetricRecordedSinceAppWasForegrounded = false
+    private var wasAppCreateCalledBeforeActivityCreate = false
+
+    fun onFenixApplicationOnCreate() {
+        wasAppCreateCalledBeforeActivityCreate = true
+    }
+
+    fun onHomeActivityOnCreate(safeIntent: SafeIntent, hasSavedInstanceState: Boolean) {
+        appOpenedAllSourceStartup.hasSavedInstance = hasSavedInstanceState
+        setAppStartupType()
+        setStartupSourceFromIntent(safeIntent, false)
+    }
+
+    fun onExternalAppBrowserOnCreate(safeIntent: SafeIntent, hasSavedInstanceState: Boolean) {
+        appOpenedAllSourceStartup.hasSavedInstance = hasSavedInstanceState
+        setAppStartupType()
+        setStartupSourceFromIntent(safeIntent, true)
+    }
+
+    fun onHomeActivityOnRestart() {
+        appOpenedAllSourceStartup.type = Type.HOT
+        appOpenedAllSourceStartup.hasSavedInstance = false
+    }
+
+    fun receivedIntentInHomeActivity(safeIntent: SafeIntent) {
```

Comment:
^ Related to the comment above above – how do the other vars get set? I need to run the code myself and relearn which methods are called when to validate the data is set correctly. Is it possible to write your code so that it's obvious to me what your assumptions are? That I can validate the code works without even running it, based on your assumptions (perhaps they need to be documented in comments)?